pkg/controller/topo/entity/reconciler.go

```diff
@@ -90,41 +89,75 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 }
 
 func (r *Reconciler) reconcileCreate(ctx context.Context, entity *v1beta1.Entity) (reconcile.Result, error) {
-	// Add the finalizer to the entity if necessary
-	if !k8s.HasFinalizer(entity, topoFinalizer) {
-		k8s.AddFinalizer(entity, topoFinalizer)
-		err := r.client.Update(ctx, entity)
-		if err != nil {
-			log.Warnf("Failed to reconcile creating entity %s, %s, %s", entity.Name, entity.Namespace, err)
+	// Get topo service name
+	topoServiceName := entity.Spec.ServiceName
+
+	// Check if topo service is available
+	topoNamespacedName := types.NamespacedName{Namespace: entity.Namespace, Name: topoServiceName}
+	topoServiceObject := &corev1.Service{}
+	if err := r.client.Get(ctx, topoNamespacedName, topoServiceObject); err != nil && k8serrors.IsNotFound(err) && entity.Status.State != v1beta1.StatePending {
+		// Set the state to StatePending if topo service is not found (deleted).
+		entity.Status = v1beta1.EntityStatus{State: v1beta1.StatePending}
+		if err := r.client.Status().Update(ctx, entity); err != nil {
+			log.Warnf("Failed to reconcile updating state of entity %s, %s, %s", entity.Name, entity.Namespace, err)
 			return reconcile.Result{}, err
 		}
-	}
-
-	// Connect to the topology service
-	conn, err := grpc.ConnectService(r.client, entity.Namespace, topoService)
-	if err != nil {
-		log.Warnf("Failed to reconcile creating entity %s, %s, %s", entity.Name, entity.Namespace, err)
+		log.Warnf("Failed to find topo service %s in namespace %s, %s", topoServiceName, entity.Namespace, err)
 		return reconcile.Result{}, err
 	}
-	defer conn.Close()
 
-	client := topo.NewTopoClient(conn)
+	switch entity.Status.State {
+	case v1beta1.StatePending:
+		entity.Status = v1beta1.EntityStatus{State: v1beta1.StateAdding}
+		err := r.client.Status().Update(ctx, entity)
+		if err != nil {
+			log.Warnf("Failed to reconcile updating state of entity %s, %s, %s", entity.Name, entity.Namespace, err)
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	case v1beta1.StateAdding:
+		// Add the finalizer to the entity if necessary
+		if !k8s.HasFinalizer(entity, topoFinalizer) {
+			k8s.AddFinalizer(entity, topoFinalizer)
+			err := r.client.Update(ctx, entity)
+			if err != nil {
+				log.Warnf("Failed to reconcile creating entity %s, %s, %s", entity.Name, entity.Namespace, err)
+				return reconcile.Result{}, err
+			}
```

Comment:
I think we should return after the finalizer is added. The update will create a new version of the entity and trigger the reconciler again so the `Adding` phase can continue.